sdk/typescript/src/bcs/index.js

```diff
@@ -0,0 +1,663 @@
+        );
+
+        MoveBCS.registerType(
+            MoveBCS.BOOL,
+            (writer, data) => writer.write8(data),
+            (reader) => reader.read8() == '1',
+            (bool) => (bool.constructor === Boolean)
+        );
+
+        MoveBCS.registerType(
+            MoveBCS.STRING,
+            (writer, data) => writer.writeVec(data, (writer, el) => writer.write8(el.charCodeAt(0))),
+            (reader) => {
+                return reader
+                    .readVec((reader) => reader.read8())
+                    .map((el) => String.fromCharCode(el))
+                    .join('');
+            },
+            (str) => (str.constructor === String) && /^[\x00-\x7F]*$/.test(str)
+        )
+    }();
+
+    /**
+     * Deserialize BCS into a JS type.
+     * 
+     * @example
+     * // use util to form an Uint8Array buffer
+     * let buffer = MoveBCS.util.fromHex('FFFFFFFF');
+     * let data = MoveBCS.de(BCS.U32, buffer);
+     * 
+     * console.assert(data == '4294967295');
+     * 
+     * @param {String} type Name of the type to deserialize (must be registered).
+     * @param {Uint8Array} data Data to deserialize.
+     * @return {Number|BN|Array<Any>|Object|Boolean} Deserialized data.
+     */
+    MoveBCS.de = function deserialize(type, data) {
+        if (!this.hasType(type)) {
+            throw new Error(`No implementation found for type ${type}`);
+        }
+
+        return this.types.get(type).decode(data);
+    };
+
+    /**
+     * Serialize data into BCS.
+     * 
+     * @example
+     * MoveBCS.registerVectorType('vector<u8>', 'u8');
+     * 
+     * let serialized = MoveBCS
+     *   .ser('vector<u8>', [1,2,3,4,5,6])
+     *   .toBytes();
+     * 
+     * console.assert(MoveBCS.util.toHex(serialized) === '06010203040506');
+     * 
+     * @param {String} type Name of the type to serialize (must be registered).
+     * @param {Uint8Array} data Data to serialize.
+     * @param {Number} [size = 1024] Serialization buffer size. Default 1024 = 1KB.
+     * @return {BcsReader} A BCS reader instance. Usually you'd want to call `.toBytes()`
+     */
+    MoveBCS.ser = function serialize(type, data, size = 1024) {
+        if (!this.hasType(type)) {
+            throw new Error(`No implementation found for type ${type}`);
+        }
+
+        return this.types.get(type).encode(data, size);
+    };
+
+    MoveBCS.util = {};
+
+    /**
+     * Turn a hex string into a Uint8Array.
+     * 
+     * @param {String} hexString A hex string.
+     * @returns {Uint8Array} A buffer to use when deserializing.
+     */
+    MoveBCS.util.fromHex = function fromHex(hexString) {
+        return new Uint8Array(hexString.match(/../g).map((h) => parseInt(h, 16)));
+    };
+
+    /**
+     * Turn Uint8Array into a hex string (lowercased).
+     * 
+     * @param {Uint8Array} buffer Uint8Array to encode as HEX.
+     * @returns {String} hex representation of BCS.
+     */
+    MoveBCS.util.toHex = function toHex(buffer) {
```

Comment:
we've needed similar functions elsewhere ([like this](https://github.com/MystenLabs/sui/blob/59297d610eb23681c8f62167203e5792d2d28fbc/explorer/client/src/pages/object-result/ObjectLoaded.tsx#L139))- we should probably have some sort of minimal shared string util library for Typescript, and standardize on one instead of shipping multiple.

@damirka how do you feel about splitting these two out ?

Reply:
I've put it under `serialization` folder in SDK.